File: Best-Time-to-Buy-and-Sell-Stock-with-Cooldown.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int hold = -prices[0];
        int sold = 0;
        int rest = 0;

        if(n == 0) return 0;

        for(int i = 1; i < n; i++){
            int prevh = hold;
            int prevs = sold;
            int prevr = rest;

            hold = Math.max(prevh, prevr - prices[i]);
            sold = prevh + prices[i];
            rest = Math.max(prevr, prevs);
        }
        return Math.max(sold, rest);
    }
}

File: Best-Time-to-Buy-and-Sell-Stock-with-Cooldown.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
        let n = prices.length;
        let hold = -prices[0];
        let sold = 0;
        let rest = 0;

        if(n == 0) return 0;

        for(let i = 1; i < n; i++){
            let prevh = hold;
            let prevs = sold;
            let prevr = rest;

            hold = Math.max(prevh, prevr - prices[i]);
            sold = prevh + prices[i];
            rest = Math.max(prevr, prevs);
        }
        return Math.max(sold, rest);
};